File: Strand/BLE/HelloSuppression.swift
import Foundation
import WhoopProtocol

/// Whether to send the WHOOP 5/MG CLIENT_HELLO at all on this connect.
///
/// The #1635 field captures show the hello is what ends the link. Across two sessions and sixteen writes it
/// was never once acknowledged, and the drop is locked to the HELLO rather than to the connect: 3.158s,
/// 3.159s, 3.155s after the write, cycle after cycle, while live HR streams happily over the standard
/// profile the whole time.
///
/// An HCI capture later established the cause: the phone transmits an SMP Pairing Request and the strap
/// answers `Pairing Not Supported` (0x05). The refusal is categorical — the encrypted bond this hello waits
/// behind can never arrive on a 5/MG — so the write is a handshake that has never once succeeded knocking a
/// strap that CAN deliver live HR off every few seconds.
///
/// Suppressing it after the give-up latches trades an unreachable capability (puffin commands, history
/// offload) for one that demonstrably works (continuous live HR) — the "Live HR, not fully paired" state
/// the app already models (#69), reached deliberately instead of never at all.
///
/// [userInitiated] always re-attempts: suppression is a fallback for automatic reconnects, never a
/// permanent verdict. Someone who puts the strap in pairing mode and presses Connect must get a fresh try.
/// That try is ONE connect's worth and nothing more — the latch itself survives the tap, so if the strap
/// still will not answer, the automatic reconnects behind it stay suppressed instead of re-earning the
/// give-up over five more refusals. The latch is dropped by a genuine bond (it demonstrably works now) or
/// by forgetting the strap, which are the two events that are actually evidence about this device.
///
/// Kotlin twin: `com.noop.ble.shouldSendClientHello`.
func shouldSendClientHello(suppressedForDevice: Bool, userInitiated: Bool) -> Bool {
    !suppressedForDevice || userInitiated
}

/// How many consecutive refusals this cause needs before the give-up latches.
///
/// The flat 5 was argued for the AUTH-REFUSAL branch and only makes sense there: the pairing hint shows at
/// 2, the hint asks the user to do something (close the official app, free a stale phone pairing), and the
/// extra cycles are the time to do it before NOOP stops hammering.
///
/// An unanswered handshake gives the user nothing to act on. The write vanishes, the strap is not refusing
/// anything it could be talked out of, and the outcome — suppress the hello and keep streaming live HR —
/// needs no permission and costs no capability that was reachable anyway. Every cycle spent waiting is a
/// ~4.8s link drop bought for nothing, so this branch stops at 3.
///
/// Not 2, which is exactly where the pairing hint fires: a strap whose hello is unanswered twice by some
/// transient would latch a PERSISTED verdict with no margin at all. 3 keeps one cycle of margin and still
/// takes roughly 40% off the churn.
///
/// Kotlin twin: `com.noop.ble.UNANSWERED_GIVE_UP_THRESHOLD`.
let unansweredGiveUpThreshold = 3

/// The give-up threshold for the refusal in hand. Keyed on the same `authRefusal` that decides whether the
/// give-up suppresses or pauses (`giveUpSuppressesHello`), so the two can never disagree about which branch
/// a refusal belongs to.
///
/// Kotlin twin: `com.noop.ble.giveUpThresholdFor`.
func giveUpThresholdFor(authRefusal: Bool, pauseThreshold: Int) -> Int {
    authRefusal ? pauseThreshold : unansweredGiveUpThreshold
}

/// Should the give-up latch suppress the hello, rather than pause auto-reconnect?
///
/// The two give-up causes want opposite treatment, which is why they are split here rather than sharing
/// one branch:
///
///  - An AUTH REFUSAL means the strap actively declined the encrypted bond — typically because it is still
///    held by the official WHOOP app. Reconnecting cannot help until the user acts, so pausing is right.
///  - An UNANSWERED HANDSHAKE means the write vanished. The link itself is healthy and streaming, so
///    pausing throws away working live HR to punish a handshake nobody is waiting on. Suppress the hello
///    and stay connected instead.
///
/// Splitting them is the whole point: before this, both ended in the same pause, so the strap that could
/// still deliver HR was treated exactly like the one that could not.
///
/// Kotlin twin: `com.noop.ble.giveUpSuppressesHello`.
func giveUpSuppressesHello(authRefusal: Bool) -> Bool { !authRefusal }

/// Does this disconnect count as the strap refusing to bond?
///
/// Two observations mean the same thing for the purpose of giving up: the strap answered the bond write
/// with INSUFFICIENT_AUTHENTICATION / INSUFFICIENT_ENCRYPTION, or it never answered the CLIENT_HELLO at
/// all. The second was invisible to the give-up on this platform, because the gate tested only the write
/// error — and an unanswered handshake produces no write error at all, just a link that drops a few
/// seconds later. The consequence is the loop #1635 describes, with nothing able to end it.
///
/// Deliberately does NOT widen what counts as an AUTH refusal — [helloUnacked] is a separate signal carried
/// separately, so the user-facing guidance can stay honest about which was observed. An auth refusal
/// supports naming a cause; an unanswered handshake does not.
///
/// Pure so the rule is unit-tested without a radio. Kotlin twin: `com.noop.ble.countsAsBondRefusal`.
func countsAsBondRefusal(
    isAuthRefusalStatus: Bool,
    helloUnacked: Bool,
    alreadyBonded: Bool,
    family: DeviceFamily
) -> Bool {
    if alreadyBonded { return false }        // bonded already — not a pairing problem
    if family != .whoop5 { return false }    // the 4.0 bonds cleanly; this is the 5/MG handshake
    return isAuthRefusalStatus || helloUnacked
}

/// UserDefaults key holding the hello-suppression latch for one strap.
///
/// Per device, and lowercased for the same reason the firmware key is: the same strap can present its
/// identifier in different cases across sessions, and a case-sensitive key would silently latch a second
/// time under a second key instead of reading the first.
///
/// DIVERGENCE FROM ANDROID (deliberate, PII): the identifier here is the CoreBluetooth-local peripheral
/// UUID, which is per-install rather than the hardware address. The key SHAPE is shared with the Kotlin
/// twin so the two are recognisably the same latch; the value that goes into it is not the same kind of
/// identifier, and deliberately so.
///
/// Kotlin twin: `com.noop.ble.helloSuppressionPrefKey`.
func helloSuppressionPrefKey(_ peripheralId: String?) -> String? {
    guard let raw = peripheralId?.trimmingCharacters(in: .whitespacesAndNewlines), !raw.isEmpty else {
        return nil
    }
    return "noop.helloUnanswered.\(raw.lowercased())"
}

/// Read/write the per-strap hello-suppression latch. Kept beside the key helper so no caller hand-rolls
/// the key, which is how the Android side ended up with a case-sensitivity bug worth a comment.
enum HelloSuppressionStore {
    static func suppressed(_ peripheralId: String?) -> Bool {
        guard let key = helloSuppressionPrefKey(peripheralId) else { return false }
        return UserDefaults.standard.bool(forKey: key)
    }

    static func setSuppressed(_ peripheralId: String?, _ suppressed: Bool) {
        guard let key = helloSuppressionPrefKey(peripheralId) else { return }
        if suppressed {
            UserDefaults.standard.set(true, forKey: key)
        } else {
            // Remove rather than store false: an absent key and a false one mean the same thing, and this
            // keeps a strap that never latched from leaving a permanent entry behind.
            UserDefaults.standard.removeObject(forKey: key)
        }
    }
}

/// May the keep-alive timer do its work on this link?
///
/// #1635: the timer used to gate on the genuine encrypted bond alone. A SUPPRESSED 5/MG never reaches
/// one, so that gate switched the whole timer off for exactly the link the suppression keeps up for
/// hours — taking the liveness watchdog with it and leaving a stalled stream with nothing to bounce it.
/// The stable-live-HR promise would have quietly depended on a stream nothing was watching.
///
/// So a 5/MG that is STREAMING (`bonded` is the live-HR shortcut of #8/#69, set when HR actually arrives
/// over the standard profile) is admitted too. Deliberately narrower than the Android twin, which admits
/// any `bonded` link: here it is 5/MG only, and only as far as the watchdog — the caller re-guards on the
/// real bond before anything puffin-framed, which an unbonded link cannot land anyway.
///
/// Pure so the rule is pinned without a radio.
func keepAliveMayRun(connected: Bool, didBond: Bool, bonded: Bool, family: DeviceFamily) -> Bool {
    guard connected else { return false }
    if didBond { return true }
    return bonded && family == .whoop5
}

File: android/app/src/main/java/com/noop/ble/HelloSuppression.kt
package com.noop.ble

/**
 * Whether to send the WHOOP 5/MG CLIENT_HELLO at all on this connect.
 *
 * The #1635 field captures show the hello is what ends the link. Across two sessions and sixteen writes it
 * was never once acknowledged, and the drop is locked to the HELLO rather than to the connect: 3.158s,
 * 3.159s, 3.155s after the write, cycle after cycle, while live HR streams happily over the standard
 * profile the whole time. The bond-state trace added for this (#1639) records no OS pairing attempt at all
 * — the write simply never completes and the local stack drops the ACL, which the log reports as an
 * unattributed GATT status 22.
 *
 * That trace was later read as "the strap is not refusing a pairing". An HCI capture disproves it: the
 * phone DOES transmit an SMP Pairing Request and the strap answers `Pairing Not Supported` (0x05). The
 * refusal is real, and it is categorical — the encrypted bond this hello was waiting behind can never
 * arrive on a 5/MG. That does not invalidate the suppression, whose evidence is about the hello's own
 * fate, but it does mean suppressing it leaves the app attempting NEITHER handshake. Hence the opt-in
 * override below.
 *
 * The result is a strap that CAN deliver live HR being knocked off every five seconds by a handshake that
 * has never once succeeded. Suppressing the hello after the give-up latches trades an unreachable
 * capability (puffin commands, history offload) for one that demonstrably works (continuous live HR) —
 * the "Live HR, not fully paired" state the app already models (#69), reached deliberately instead of
 * never at all.
 *
 * [userInitiated] always re-attempts: suppression is a fallback for automatic reconnects, never a
 * permanent verdict. Someone who puts the strap in pairing mode and presses Connect must get a fresh try.
 * That try is ONE connect's worth and nothing more — the latch itself survives the tap, so if the strap
 * still will not answer, the automatic reconnects behind it stay suppressed instead of re-earning the
 * give-up over five more refusals. The latch is dropped by a genuine bond (it demonstrably works now) or
 * by forgetting the strap, which are the two events that are actually evidence about this device.
 *
 * Deliberately NOT re-armed by "an OS pairing exists". That looks right — a pairing is new evidence the
 * handshake might work — but the condition never goes away, so a strap that pairs and STILL will not
 * answer would have the latch bypassed on every connect for good: hello, drop at ~4.8s, reconnect,
 * forever, with the give-up powerless to stop it. That is the unbounded loop this suppression exists to
 * end, reintroduced through the back door. The explicit-bond experiment instead clears the latch ONCE at
 * the moment it asks for a pairing, which is self-limiting: the next failure re-latches and the condition
 * to clear it again does not recur.
 */
internal fun shouldSendClientHello(
    suppressedForDevice: Boolean,
    userInitiated: Boolean,
    overrideSuppression: Boolean = false,
): Boolean = !suppressedForDevice || userInitiated || overrideSuppression

/**
 * How many times the #1635 override may write an unanswered hello before it stops on its own.
 *
 * The override needs a bound that does not depend on the disconnect status, and the shared give-up cannot
 * supply one. `shouldCountNeverBondedSelfDrop` excludes `GATT_CONN_TERMINATE_LOCAL_HOST` (0x16) because
 * that normally means WE hung up — our own `gatt.disconnect()` or the bond-watchdog bounce — and counting
 * those would be self-referential. But the hello failure arrives as exactly that status: the strap answers
 * the write with ATT `Insufficient Authentication`, the local stack tries to elevate security, SMP is
 * refused, and the stack tears the ACL down. Not our teardown, same status code.
 *
 * Field result of leaving it unbounded: 57 reconnect cycles in an hour, each ~4.8s, with nothing able to
 * stop it. So the cap lives here instead — small enough to spare the battery, large enough that a strap
 * which answers on a later attempt is not written off.
 */
internal const val HELLO_OVERRIDE_MAX_ATTEMPTS = 6

/**
 * May the override write another hello, given how many it has already written unanswered?
 *
 * Counted per app process and reset on a genuine bond. A process restart resets it too, which is the
 * honest limit of an in-memory bound — but the loop it exists to stop runs within one process, and a
 * restart is not the failure mode.
 */
internal fun overrideHelloStillAllowed(
    attemptsSoFar: Int,
    cap: Int = HELLO_OVERRIDE_MAX_ATTEMPTS,
): Boolean = attemptsSoFar < cap

/**
 * Is the override actually in force — opted in AND with budget left?
 *
 * Every reader must ask through here rather than reading the pref directly. A spent override is not
 * "on": the hello genuinely stops going out, and a caller still reading the raw pref believes a hello
 * is on the wire that is not. That is the `didBond`-reader trap from CLAUDE.md pointed the other way,
 * and it had already reached [shouldCountNeverBondedSelfDrop]: with the budget spent, the app is in the
 * deliberate suppressed-hello live-HR state, but the raw pref reported the hello as still being sent, so
 * the never-bonded detector kept counting self-drops and would eventually pause auto-reconnect and raise
 * the stale-pairing guide — telling the user to re-pair for a cause that never happened.
 *
 * The boundary is deliberate: the link carrying the LAST permitted hello reads as inactive at its own
 * disconnect, because [helloOverrideAttempts] is charged when that hello is written. Under-counting that
 * one drop is the wanted behaviour — no further hello follows it, so the loop is already over, and
 * counting it could trip the pause at the exact moment the override retires.
 */
internal fun helloOverrideActive(
    optedIn: Boolean,
    attemptsSoFar: Int,
    cap: Int = HELLO_OVERRIDE_MAX_ATTEMPTS,
): Boolean = optedIn && overrideHelloStillAllowed(attemptsSoFar, cap)

/**
 * Does flipping the switch to on re-arm a spent budget?
 *
 * Turning the experiment off and on again is the user explicitly asking for another try, and without
 * this they would not get one: the counter outlives the toggle, so re-enabling a spent override does
 * nothing whatsoever — and silently, because the give-up line is one-shot and has already latched. The
 * edge is sampled per connect, which is the only moment the override can act anyway.
 */
internal fun helloOverrideBudgetRearms(optedInNow: Boolean, optedInLastSeen: Boolean): Boolean =
    optedInNow && !optedInLastSeen

/**
 * The line printed when the override gives up, so the log says why the hello stopped rather than leaving
 * a reader to notice its absence.
 */
internal fun helloOverrideExhaustedLine(attempts: Int): String =
    "WHOOP 5/MG: \"send hello despite bond refusal\" has written $attempts unanswered hellos — stopping." +
        " The strap answers the write with Insufficient Authentication and refuses SMP pairing, so the" +
        " handshake cannot complete; continuing would only loop the link. Turn the switch off to return to" +
        " the stable live-HR state (#1635)."

/**
 * Should an explicit Connect keep the link it already holds, instead of rebuilding it?
 *
 * `getConnectedWhoopDevice()` answers "the OS has this device connected", which is true of our own live
 * GATT too — so the Easy-connect path re-attached over a link that was already working. Field log
 * 260901-0121: a link up 18m 32s and streaming was replaced on a tap.
 *
 * Every clause is load-bearing:
 *  - [genuinelyBonded]: the ONLY state where a reconnect can achieve nothing. A suppressed strap is
 *    deliberately excluded — there the tap IS the handshake retry, and the hello can only be written on a
 *    new link's discovery, so it has to reconnect.
 *  - [sameDevice]: a tap meant for a different strap must not be swallowed by the one in hand.
 *  - [silentMs] < [stallFuseMs]: `connected` alone is not evidence the link WORKS. A silently dead GATT
 *    reports connected until the watchdog bounces it, and that is exactly when someone taps Connect —
 *    keeping the link there would make the button inert in the one state it exists for. The fuse is the
 *    watchdog's own, so the two can never disagree about whether this link is alive.
 */
internal fun connectKeepsExistingLink(
    genuinelyBonded: Boolean,
    connected: Boolean,
    sameDevice: Boolean,
    silentMs: Long,
    stallFuseMs: Long,
): Boolean = genuinelyBonded && connected && sameDevice && silentMs < stallFuseMs

/**
 * How many consecutive refusals this cause needs before the give-up latches.
 *
 * The flat 5 was argued for the AUTH-REFUSAL branch and only makes sense there: the pairing hint shows at
 * 2, the hint asks the user to do something (close the official app, free a stale phone pairing), and the
 * extra cycles are the time to do it before NOOP stops hammering.
 *
 * An unanswered handshake gives the user nothing to act on. The write vanishes, the strap is not refusing
 * anything it could be talked out of, and the outcome — suppress the hello and keep streaming live HR —
 * needs no permission and costs no capability that was reachable anyway. Every cycle spent waiting is a
 * ~4.8s link drop bought for nothing, so this branch stops at 3.
 *
 * Not 2, which is exactly where the pairing hint fires: a strap whose hello is unanswered twice by some
 * transient would latch a PERSISTED verdict with no margin at all. 3 keeps one cycle of margin and still
 * takes roughly 40% off the churn.
 */
internal const val UNANSWERED_GIVE_UP_THRESHOLD = 3

/**
 * The give-up threshold for the refusal in hand. Keyed on the same [authRefusal] that decides whether the
 * give-up suppresses or pauses ([giveUpSuppressesHello]), so the two can never disagree about which branch
 * a refusal belongs to.
 */
internal fun giveUpThresholdFor(authRefusal: Boolean, pauseThreshold: Int): Int =
    if (authRefusal) pauseThreshold else UNANSWERED_GIVE_UP_THRESHOLD

/**
 * May a pairing-hint clear also drop the PERSISTED hello-suppression latch?
 *
 * Only a genuine bond. A bond is proof the handshake works on this strap NOW, so the old verdict is stale
 * and must go. A user Connect is not proof of anything: it already gets its one fresh attempt from the
 * retry flag ([shouldSendClientHello]'s `userInitiated`), and dropping the latch on top of that un-
 * suppressed every AUTOMATIC reconnect behind the attempt — so a single tap cost five more refusals and
 * ~55s of link churn to re-earn a verdict about firmware that had not changed. Field log 260901-0121:
 * latched 01:02:31, stable for 18 minutes, one tap at 01:21:09, straight back into the loop.
 *
 * Forgetting the strap clears it too, but writes the pref directly — the strap is being released, so there
 * is no hint left to clear and nothing to route through here.
 *
 * Apple has only ever cleared the latch on a genuine bond or a forget; this predicate is what makes the
 * Android side say the same thing.
 */
internal fun pairingHintClearDropsSuppressionLatch(genuineBond: Boolean): Boolean = genuineBond

/**
 * Should the give-up latch suppress the hello, rather than pause auto-reconnect?
 *
 * The two give-up causes want opposite treatment, which is why they are split here rather than sharing
 * one branch:
 *
 *  - An AUTH REFUSAL means the strap actively declined the encrypted bond — typically because it is still
 *    held by the official WHOOP app. Reconnecting cannot help until the user acts, so pausing is right.
 *  - An UNANSWERED HANDSHAKE means the write vanished. The link itself is healthy and streaming, so
 *    pausing throws away working live HR to punish a handshake nobody is waiting on. Suppress the hello
 *    and stay connected instead.
 *
 * Splitting them is the whole point: before this, both ended in the same pause, so the strap that could
 * still deliver HR was treated exactly like the one that could not.
 */
internal fun giveUpSuppressesHello(authRefusal: Boolean): Boolean = !authRefusal

/**
 * SharedPreferences key holding the hello-suppression latch for one device.
 *
 * Per device, and lowercased for the same reason [firmwarePrefKey] is: the same strap can present its
 * address in different cases across sessions, and a case-sensitive key would silently latch a second time
 * under a second key instead of reading the first.
 */
internal fun helloSuppressionPrefKey(peripheralId: String?): String? =
    peripheralId?.trim()?.takeIf { it.isNotEmpty() }?.let { "noop.helloUnanswered.${it.lowercase()}" }
